### PR TITLE
Rs 160 sqs connector

### PR DESCRIPTION
**Pull request by lagache**

### CODE REVIEW

File: doc/AWSSQSConnector.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+```
+
+_Handler inputs:_
+
+```typescript
+(event: MESSAGE, app: Reshuffle) => void
+
+interface MESSAGE {
+  /**
+   * A unique identifier for the message. A MessageIdis considered unique across all AWS accounts for an extended period of time.
```
Comment (duckranger) [inline doc/AWSSQSConnector.md:83]: A MessageIdis  => A MessageId is

File: doc/AWSSQSConnector.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+   * An MD5 digest of the non-URL-encoded message body string.
+   */
+  MD5OfBody?: string
+  /**
+   * The message's contents (not URL-encoded).
+   */
+  Body?: string
+  /**
+   * A map of the attributes requested in  ReceiveMessage  to their respective values. Supported attributes:    ApproximateReceiveCount     ApproximateFirstReceiveTimestamp     MessageDeduplicationId     MessageGroupId     SenderId     SentTimestamp     SequenceNumber     ApproximateFirstReceiveTimestamp and SentTimestamp are each returned as an integer representing the epoch time in milliseconds.
+   */
```
Comment (duckranger) [inline doc/AWSSQSConnector.md:100]: Consider line breaks here?

File: doc/AWSSQSConnector.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+  Attributes?: MessageSystemAttributeMap
+  /**
+   * An MD5 digest of the non-URL-encoded message attribute string. You can use this attribute to verify that Amazon SQS received the message correctly. Amazon SQS URL-decodes the message before creating the MD5 digest. For information about MD5, see RFC1321.
+   */
+  MD5OfMessageAttributes?: string
+  /**
+   * Each message attribute consists of a Name, Type, and Value. For more information, see Amazon SQS Message Attributes in the Amazon Simple Queue Service Developer Guide.
```
Comment (duckranger) [inline doc/AWSSQSConnector.md:107]: Link please?

File: doc/AWSSQSConnector.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+  /**
+   * The URL of the Amazon SQS queue to which a message is sent. Queue URLs and names are case-sensitive.
+   */
+  QueueUrl: string
+  /**
+   * The message to send. The minimum size is one character. The maximum size is 256 KB.  A message can include only XML, JSON, and unformatted text. The following Unicode characters are allowed:  #x9 | #xA | #xD | #x20 to #xD7FF | #xE000 to #xFFFD | #x10000 to #x10FFFF  Any characters not included in this list will be rejected. For more information, see the W3C specification for characters.
+   */
+  MessageBody: string
+  /**
+   *  The length of time, in seconds, for which to delay a specific message. Valid values: 0 to 900. Maximum: 15 minutes. Messages with a positive DelaySeconds value become available for processing after the delay period is finished. If you don't specify a value, the default value for the queue applies.   When you set FifoQueue, you can't set DelaySeconds per message. You can set this parameter only on a queue level.
```
Comment (duckranger) [inline doc/AWSSQSConnector.md:150]: `applies.   When` - double space?

File: doc/AWSSQSConnector.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+```ts
+(params: AWS.SQS.Types.SendMessageRequest) => Promise<PromiseResult<AWS.SQS.SendMessageResponse, AWS.AWSError>>
+
+export interface SendMessageRequest {
+  /**
+   * The URL of the Amazon SQS queue to which a message is sent. Queue URLs and names are case-sensitive.
+   */
+  QueueUrl: string
+  /**
+   * The message to send. The minimum size is one character. The maximum size is 256 KB.  A message can include only XML, JSON, and unformatted text. The following Unicode characters are allowed:  #x9 | #xA | #xD | #x20 to #xD7FF | #xE000 to #xFFFD | #x10000 to #x10FFFF  Any characters not included in this list will be rejected. For more information, see the W3C specification for characters.
```
Comment (duckranger) [inline doc/AWSSQSConnector.md:146]: `are allowed:  #x9 | ` - double space?

File: doc/AWSSQSConnector.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+   */
+  MessageBody: string
+  /**
+   *  The length of time, in seconds, for which to delay a specific message. Valid values: 0 to 900. Maximum: 15 minutes. Messages with a positive DelaySeconds value become available for processing after the delay period is finished. If you don't specify a value, the default value for the queue applies.   When you set FifoQueue, you can't set DelaySeconds per message. You can set this parameter only on a queue level.
+   */
+  DelaySeconds?: Integer
+  /**
+   * Each message attribute consists of a Name, Type, and Value. For more information, see Amazon SQS Message Attributes in the Amazon Simple Queue Service Developer Guide.
```
Comment (duckranger) [inline doc/AWSSQSConnector.md:154]: link please?

File: doc/AWSSQSConnector.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+   */
+  MessageAttributes?: MessageBodyAttributeMap
+  /**
+   * The message system attribute to send. Each message system attribute consists of a Name, Type, and Value.    Currently, the only supported message system attribute is AWSTraceHeader. Its type must be String and its value must be a correctly formatted AWS X-Ray trace header string.   The size of a message system attribute doesn't count towards the total size of a message.
```
Comment (duckranger) [inline doc/AWSSQSConnector.md:158]: `header string.   The s` - double space?

File: doc/AWSSQSConnector.md [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
... within the deduplication interval are treated as duplicates and only one copy of the message is delivered.   If you send one message with ContentBasedDeduplication enabled and then another message with a MessageDeduplicationId that is the same as the one generated for the first MessageDeduplicationId, the two messages are treated as duplicates and only one copy of the message is delivered.     The MessageDeduplicationId is available to the consumer of the message (this can be useful for troubleshooting delivery issues). If a message is sent successfully but the acknowledgement is lost and the message is resent with the same MessageDeduplicationId after the deduplication interval, Amazon SQS can't detect duplicate messages. Amazon SQS continues to keep track of the message deduplication ID even after the message is received and deleted.  The maximum length of MessageDeduplicationId is 128 characters. MessageDeduplicationId can contain alphanumeric characters (a-z, A-Z, 0-9) and punctuation (!"#$%&amp;'()*+,-./:;&lt;=&gt;?@[\]^_`{|}~). For best practices of using MessageDeduplicationId, see Using the MessageDeduplicationId Property in the Amazon Simple Queue Service Developer Guide.
```
Comment (duckranger) [inline doc/AWSSQSConnector.md:162]: Generally speaking, for both this and other long-form comments, I think we should have a one lineer and point to the relevant URL on AWS docs (e.g. https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_SendMessage.html)

File: doc/AWSSQSConnector.md [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+}
+const response = await sqsConnector.sendMessage(params)
+console.log(response.MessageId)
+```
+
+#### SDK
+
+##### <a name="sdk"></a>SDK action
+
+Returns a AWS SQS client instance
```
Comment (duckranger) [inline doc/AWSSQSConnector.md:209]: an

File: doc/AWSSQSConnector.md [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+// Full list of available actions in SDK: https://github.com/aws/aws-sdk-js/blob/master/clients/sqs.d.ts
```
Comment (duckranger) [inline doc/AWSSQSConnector.md:210]: Maybe make this url clickable?

File: doc/AWSSQSConnector.md [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+
+_Example:_
+
+Receive messages from queue
+```js
+const response = await sqsConnector.sdk().receiveMessage({ QueueUrl }).promise()
+console.log(response.Messages)
+```
+
+For more examples, go to [here](https://github.com/reshufflehq/reshuffle/tree/master/examples/aws/sqs-messages.js).
```
Comment (duckranger) [inline doc/AWSSQSConnector.md:232]: don't need the word `to`

File: src/AWSSQSConnector.ts [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    handler: CoreEventHandler,
+    eventId?: string,
+  ): EventConfiguration {
+    const eid = eventId || `AWS/SQS/message/${options.queueUrl}/${this.id}`
+    return this.eventManager.addEvent(options, handler, eid)
+  }
+
+  protected async onInterval(): Promise<void> {
+    const events = Object.values(this.eventManager.eventConfigurationSet)
+    const queueUrls = events.reduce<Record<string, AWSSQSConnectorEventOptions>>(
```
[the remark this answers is not in the document]
Comment (lagache) [reply src/AWSSQSConnector.ts:31]: yeah it was to make the obj simpler but I''ll change it back

File: src/AWSSQSConnector.ts [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+  private async getQueueMessages(QueueUrl: string, deleteAfterReceive = true) {
+    const res = await this.client.receiveMessage({ QueueUrl }).promise()
+    if (res.Messages && res.Messages.length) {
+      this.app
+        .getLogger()
+        .info(
+          `Reshuffle - AWSSQSConnector: ${res.Messages.length} message(s) received from queue ${QueueUrl}`,
+        )
+
+      if (deleteAfterReceive && res.Messages) {
```
[the remark this answers is not in the document]
Comment (lagache) [reply src/AWSSQSConnector.ts:55]: true

File: doc/AWSSQSConnector.md (hunk 7 above)
Comment (lagache) [reply doc/AWSSQSConnector.md:158]: that's copied from AWS doc, but I've changed it

File: doc/AWSSQSConnector.md (hunk 8 above)
Comment (lagache) [reply doc/AWSSQSConnector.md:162]: I agree

**Pull request merged**